### PR TITLE
ARROW-3986: [C++] Write prose documentation

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@wesm @fsaintjacques

File: cpp/src/arrow/status.h [hunk 1]
```diff
@@ -135,45 +145,54 @@ class ARROW_EXPORT Status {
   Status& operator&=(const Status& s) noexcept;
   Status& operator&=(Status&& s) noexcept;
 
-  // Return a success status.
+  /// Return a success status
   static Status OK() { return Status(); }
 
-  // Return a success status with extra info
+  /// Return a success status with extra info
```
Comment (fsaintjacques) [inline cpp/src/arrow/status.h]: "with a caller defined message"

File: cpp/src/arrow/status.h [hunk 2]
```diff
@@ -29,6 +29,7 @@
 
 #ifdef ARROW_EXTRA_ERROR_CONTEXT
 
+/// @brief Propagate any non-successful Status to the caller
```
Comment (fsaintjacques) [inline cpp/src/arrow/status.h]: Inconsistency between `@brief` or `\brief`, and no brief (see later functions).

File: cpp/src/arrow/status.h [hunk 3]
```diff
@@ -211,34 +229,42 @@ class ARROW_EXPORT Status {
...
+  /// Return true iff the status indicates a container reaching capacity limits.
   bool IsCapacityError() const { return code() == StatusCode::CapacityError; }
+  /// Return true iff the status indicates a type error.
   bool IsTypeError() const { return code() == StatusCode::TypeError; }
+  /// Return true iff the status indicates an unknown error.
   bool IsUnknownError() const { return code() == StatusCode::UnknownError; }
+  /// Return true iff the status indicates an unimplemented operation.
   bool IsNotImplemented() const { return code() == StatusCode::NotImplemented; }
-  // An object could not be serialized or deserialized.
+  /// Return true iff the status indicates a (de) serialization failure?
```
Comment (fsaintjacques) [inline cpp/src/arrow/status.h]: Are you asking the reader? :)

File: docs/source/cpp/overview.rst [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+arrays into a longer logical sequence of values.
+
+The two-dimensional layer
+-------------------------
+
+**Schemas** describe a logical collection of several pieces of data,
+each with a distinct name and type, and optional metadata.
+
+**Columns** are like chunked arrays, but with optional metadata.
+
```
Comment (fsaintjacques) [inline docs/source/cpp/overview.rst]: Missing record batches.

File: docs/source/cpp/datatypes.rst [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+1. Using a :class:`arrow::DataType` instance (e.g. as a function argument)
+2. Using a :class:`arrow::DataType` concrete subclass (e.g. as a template
+   parameter)
+3. Using a :type:`arrow::Type::type` enum value (e.g. as the condition of
+   a switch statement)
+
+The first form (using a :class:`arrow::DataType` instance) is the most idiomatic
+and flexible.  Non-trivial types can _only_ be represented with a DataType
+instance as they are parametric.  For example, a :class:`arrow::TimestampType`
```
Comment (fsaintjacques) [inline docs/source/cpp/datatypes.rst]: > as the parameter(s) are often unknown at compile time

**Review by fsaintjacques — COMMENTED** [review]
Some minor nitpick, otherwise this is good.

**Comment by pitrou** [comment]
Thanks. I think I've addressed your comments now.

File: docs/source/cpp/overview.rst [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+**Tables** are collections of columns in accordance to a schema.  They are
+the most capable dataset-providing abstraction in Arrow.
+
+**Record batches** are collections of contiguous arrays, described
```
Comment (fsaintjacques) [inline docs/source/cpp/overview.rst]: Another important fact about record batches, they're the same size. This is critical for the future execution engine.

**Comment by pitrou** [comment]
I've pushed a couple additions here. If no objections, I may merge this soon and continue in other PRs.

**Comment by wesm** [comment]
I will have a look through today if possible and leave any comments to be addressed in follow up PRs

**Pull request closed without merging**